File: patterns/hash_tables/key_value/sort_by_frequency.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

string sort_by_frequency(const string &input_data)
{
    // Stage 1: KV counting part
    unordered_map<char, int> counter;
    for (char c : input_data)
    {
        counter[c]++;
    }

    // Stage 2: VK, basically do the inversion and count the by freq groups
    vector<vector<char>> frequency_list(input_data.size() + 1);
    for (const auto &[symbol, frequency] : counter)
    {
        frequency_list[frequency].push_back(symbol);
    }

    // Stage 3: Creating the result from the frequncy list
    string result;
    result.reserve(input_data.size());
    for (int i = frequency_list.size() - 1; i > 0; i--)
    {
        for (char c : frequency_list[i])
        {
            result.append(i, c);
        }
    }

    return result;
}

int main()
{
    string input_data = "BABBCBC";
    cout << "Input data: " << input_data << endl;
    string result = sort_by_frequency(input_data);
    cout << "Result: " << result << endl;

    return 0;
}

File: patterns/hash_tables/key_value/sort_by_frequency.py
"""
Given a list with characters. Yous should sort its symbols by their frequency,
from the most frequent to the least frequent.

e.g.
chars = ["B", "A", "B", "B", "C", "B", "C"]
Solution = BBBBCCA

"""


def sort_by_frequency(chars: list[str]) -> str:
    """This approach shoudl only use th ehash tables key value pattern"""

    # First pass, counting the frequency of each char
    counter: dict[str, int] = {}
    for char in chars:
        counter[char] = counter.get(char, 0) + 1

    # Second pass, we create the frequency list
    frequency_list: list[list[str]] = [[] for _ in range(len(chars) + 1)]

    for char, freq in counter.items():
        frequency_list[freq].append(char)

    # Create the final responmse list with the sorted chars
    result: list[str] = []
    for i in range(len(frequency_list) - 1, 0, -1):
        for char in frequency_list[i]:
            result.append(char * i)

    return "".join(result)


if __name__ == "__main__":
    input_list: list[str] = ["B", "A", "B", "B", "C", "B", "C"]

    result: str = sort_by_frequency(input_list)
    print(f"Input: {input_list} | Solution: {result}")
